flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/OnMainThreadJobManagerRunnerRegistry.java

```diff
@@ -84,9 +84,10 @@ public Collection<JobManagerRunner> getJobManagerRunners() {
     }
 
     @Override
-    public CompletableFuture<Void> localCleanupAsync(JobID jobId, Executor executor) {
+    public CompletableFuture<Void> localCleanupAsync(
+            JobID jobId, Executor cleanupExecutor, Executor executor) {
```

Comment:
> I'm wondering whether we should even use ComponentMainThreadExecutor as a type for the main thread executor to make things clear.

I tried this but not sure why but the tests got stuck.

Reply:
You have to be more specific here. Where did the tests got stuck?

Reply:
`testResourceCleanupUnderLeadershipChange` is one such test. I suspect this is happening because we busy wait here:
https://github.com/apache/flink/blob/baa67f465b77c9b3131d8153b52d2a94e3e00ef9/flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java#L238-L239.

We have a `waitUntilCondition` on the main thread which expects the cleanup which is also done by the main thread's executor. Strangely this is not an issue when we use `Executor` as the type. 🤔

Reply:
Do you have any thoughts?